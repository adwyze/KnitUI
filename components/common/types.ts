export interface IStyled<PropType> {
  customProps: PropType
  theme: any
  [htmlProp: string]: any
}
