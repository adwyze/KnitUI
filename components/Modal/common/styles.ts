import * as theme from "../../styles/variables"

export const border = `1px solid ${theme.shades.gray85}`
export const borderRadius = "0.4rem"
export const padding = {
  vertical: "2.1rem",
  horizontal: "2.8rem",
}
