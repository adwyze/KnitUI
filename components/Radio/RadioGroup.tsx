import React, {
  SFC,
  ReactNode,
  CSSProperties,
  Component,
  cloneElement,
} from "react"
import _ from "lodash"
import { RadioProps } from "./Radio"

export interface RadioChangeEventTarget extends RadioProps {
  checked: boolean
}

type size = "small" | "medium"

export interface RadioChangeEvent {
  target: RadioChangeEventTarget
  stopPropagation: () => void
  preventDefault: () => void
  nativeEvent: MouseEvent
}

interface RadioGroupProps {
  /** Value of the radio group */
  value?: any
  /** Custom styles for each radio button */
  radioStyle?: CSSProperties
  /** Function to call when radio is changed */
  onChange?: (e: RadioChangeEvent) => void
  /** default value of the group */
  defaultValue?: any
  /** Size of the radio items */
  size?: size
  /** children nodes */
  children?: ReactNode
  /** Whether the group is disabled */
  disabled?: boolean
}

const renderRadio = (props, value, setValue) => {
  const propsToPass = _.omit(props, ["children"])
  let Radios = props.children || <></>
  if (!Array.isArray(Radios)) Radios = [Radios]
  return Radios.map((Radio, index) => {
    return cloneElement(Radio, {
      groupValue: value,
      setValue,
      key: index,
      ...propsToPass,
    })
  })
}

export default class RadioGroup extends Component<RadioGroupProps, any> {
  static Item: React.FunctionComponent<RadioProps>

  static defaultProps = {
    size: "medium",
  }

  state = {
    value: this.props.value || this.props.defaultValue,
  }

  setValue = value => {
    this.setState({ value })
  }
  render() {
    return <div>{renderRadio(this.props, this.state.value, this.setValue)}</div>
  }
}
