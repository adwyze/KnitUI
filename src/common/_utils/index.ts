import { ThemedStyledFunction } from "styled-components"
import chroma from "chroma-js"
import { ColorPreset, CustomColor } from "../types"

export const insertIf = (
  obj: any = {},
  condition: boolean,
  isArray: boolean = false
) => {
  if (condition) {
    return obj
  }
  if (isArray) {
    return []
  }
  return {}
}

// https://github.com/styled-components/styled-components/issues/630
export function withProps<U>() {
  return <
    P extends keyof JSX.IntrinsicElements | React.ComponentType<any>,
    T extends object,
    O extends object = {}
  >(
    fn: ThemedStyledFunction<P, T, O>
  ): ThemedStyledFunction<P & U, T, O & U> =>
    (fn as unknown) as ThemedStyledFunction<P & U, T, O & U>
}

// Color utils

/**
 * Checks if the color is light or darks based on its L value in
 * the HSL color scale.
 * @param color A chroma color instance
 */
export const isLightColor = color => color.get("hsl.l") >= 0.5

/**
 * Returns font color of updated lightness on hover
 * @param color A chroma color instance
 */
export const getHoverFontColor = color => {
  return isLightColor(color)
    ? color.set("hsl.l", "-0.3")
    : color.set("hsl.l", "+0.3")
}

/**
 * Returns background color of updated lightness on hover
 * @param color A chroma color instance
 */
export const getHoverBackgroundColor = color => {
  return isLightColor(color)
    ? color.set("hsl.l", "-0.1")
    : color.set("hsl.l", "+0.1")
}


/**
 * Determines the appropriate font color based on the background color.
 */
export const getFontColor = (theme, backGroundColor) => {
  return isLightColor(backGroundColor)
    ? theme.chromaPalette.Neutral90
    : theme.chromaPalette.Neutral0
}

const createParsedColorTheme = (theme, backgroundColor) => ({
  background: backgroundColor,
  font: getFontColor(theme, backgroundColor),
})

/**
 * Creates an object with background and font properties initialized
 * with appropriate chroma colors.
 * @param backGroundColor A plain CSS color string
 */
export const parseCustomColor = (theme, customColor: CustomColor) => {
  if (
    !(
      chroma.valid(customColor) ||
      (typeof customColor === "object" &&
        chroma.valid(customColor.color) &&
        chroma.valid(customColor.secondaryColor))
    )
  ) {
    throw new Error("Provided color theme contains invalid color values")
  }
  if (typeof customColor === "object") {
    return {
      font: chroma(customColor.secondaryColor),
      background: chroma(customColor.color),
    }
  }
  customColor = chroma(customColor)
  return createParsedColorTheme(theme, customColor)
}

/**
 * Creates an object with background and font properties initialized
 * with appropriate chroma colors.
 * @param backGroundColor One of the predefined preset values
 */
export const parseColorPreset = (theme, backgroundColor: ColorPreset) => {
  backgroundColor = theme.secondaryPalette[backgroundColor]
  return createParsedColorTheme(theme, backgroundColor)
}

/**
 * Get Color from Styled Component Props
 */
export const getThemeColor = (props: any, color: string) => {
  const {
    theme: {
      knitui: { chromaPalette },
    },
  } = props

  return chromaPalette[color].hex()
}

/**
 * Get OS of current platform
 */
export const getOSName = () => {
  const platform = window.navigator.platform
  const userAgent = window.navigator.userAgent

  const macOSList = ["Macintosh", "MacIntel", "MacPPC", "Mac68K"]
  const windowOSList = ["Windows", "Win16", "Win32", "WinCE"]
  const iOSList = ["iPhone", "iPad", "iPod"]

  let os: string | undefined

  if (macOSList.includes(platform)) {
    os = "MacOS"
  } else if (windowOSList.includes(platform)) {
    os = "Windows"
  } else if (iOSList.includes(platform)) {
    os = "iOS"
  } else if (/Android/.test(userAgent)) {
    os = "Android"
  } else if (/Linux/.test(userAgent)) {
    os = "Linux"
  }

  return os
}
