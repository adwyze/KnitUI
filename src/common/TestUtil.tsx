import React, { ReactElement } from "react"
import { render } from "react-testing-library"
import { ThemeProvider } from "./styles"

const ProviderWrapper = ({
  children,
}: {
  children: ReactElement
}): ReactElement => {
  // Nest all providers here
  return <ThemeProvider>{children}</ThemeProvider>
}

const customRender = (ui: ReactElement, options?: object) =>
  render(ui, { wrapper: ProviderWrapper, ...options })

// re-export everything
export * from "react-testing-library"
// override render method
export { customRender as render }
