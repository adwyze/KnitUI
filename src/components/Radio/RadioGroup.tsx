import React, { useState, cloneElement } from "react"
import _ from "lodash"
import { RadioGroupProps, RadioProps } from "./Interface"
import Radio from "./Radio"

const renderRadio = (props, value, setValue) => {
  const propsToPass = _.omit(props, ["children", "defaultValue"])
  let Radios = props.children || <></>
  if (!Array.isArray(Radios)) Radios = React.Children.toArray(Radios)
  return Radios.map((radio, index) => {
    return cloneElement(radio, {
      setValue,
      style: props.style,
      key: index,
      checked: value === radio.props.value,
      ...propsToPass,
    })
  })
}

const getDefaultValue = radios => {
  const child = React.Children.count(radios)
  if (child < 2) {
    // Optionally throw error if need be
    console.error("Add 2 or more Radio Items to the Group")
  }
  switch (true) {
    case child > 1:
      const len = radios.length - 1
      for (const radio in radios) {
        if (radios[len - Number(radio)].props.checked) {
          return radios[len - Number(radio)].props.value
        }
      }
      return null
    case child === 1:
      return radios.props.checked ? radios.props.value : null
    default:
      return null
  }
}

const RadioGroup: React.FC<RadioGroupProps> & {
  Item: React.FC<RadioProps>
} = props => {
  const radioValue = getDefaultValue(props.children)
  const [value, setValue] = useState(radioValue || props.defaultValue)

  return <>{renderRadio(props, value, setValue)}</>
}

RadioGroup.defaultProps = {
  size: "medium",
}

RadioGroup.Item = Radio

export default RadioGroup
