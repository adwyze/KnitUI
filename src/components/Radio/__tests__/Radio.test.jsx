import React from "react"
import "jest-dom/extend-expect"

import { cleanup, fireEvent, render } from "../../../common/TestUtil"
import RadioGroup from "../index"

afterEach(cleanup)

describe("Radio", () => {
  describe("renders radio correctly", () => {
    it("Small with 2 in group", () => {
      const { asFragment, container, getByText } = render(
        <RadioGroup size="small">
          <RadioGroup.Item value={1}>Option 1</RadioGroup.Item>
          <RadioGroup.Item value={2}>Option 2</RadioGroup.Item>
        </RadioGroup>
      )
      expect(container.querySelector(".knit-radio-inner")).toHaveStyle(
        `height: 1.2rem`
      )
      const allRadioInputs = container.querySelectorAll(".knit-radio-input")
      fireEvent.click(allRadioInputs[1])
      expect(allRadioInputs[0]).not.toBeChecked()
      expect(allRadioInputs[1]).toBeChecked()
      expect(asFragment()).toMatchSnapshot()
    })

    it("Small with 2 in group with disabled", () => {
      const { asFragment, container } = render(
        <RadioGroup size="small" disabled defaultValue={1}>
          <RadioGroup.Item value={1}>Option 1</RadioGroup.Item>
          <RadioGroup.Item value={2}>Option 2</RadioGroup.Item>
        </RadioGroup>
      )
      const allRadioInputs = container.querySelectorAll(".knit-radio-input")
      // make sure that the inputs are disabled
      expect(allRadioInputs[0]).toHaveAttribute("disabled")
      fireEvent.click(allRadioInputs[1])
      expect(allRadioInputs[0]).toBeChecked()
      expect(allRadioInputs[1]).not.toBeChecked()
      expect(asFragment()).toMatchSnapshot()
    })

    it("Medium with 2 in a group", () => {
      const { asFragment, container } = render(
        <RadioGroup size="medium" disabled defaultValue={1}>
          <RadioGroup.Item value={1}>Option 1</RadioGroup.Item>
          <RadioGroup.Item value={2}>Option 2</RadioGroup.Item>
        </RadioGroup>
      )

      expect(container.querySelector(".knit-radio-inner")).toHaveStyle(
        `height: 1.4rem`
      )
      expect(asFragment()).toMatchSnapshot()
    })

    it("renders with single child correctly", () => {
      const { asFragment } = render(
        <RadioGroup>
          <RadioGroup.Item value={1} />
        </RadioGroup>
      )
      expect(asFragment()).toMatchSnapshot()
    })

    it("renders with no child correctly", () => {
      const { asFragment } = render(<RadioGroup />)
      expect(asFragment()).toMatchSnapshot()
    })

    it("should call the provided onClick function", () => {
      const onClick = jest.fn()
      const { container } = render(
        <RadioGroup>
          <RadioGroup.Item onClick={onClick} value={1}>
            Option 1
          </RadioGroup.Item>
          <RadioGroup.Item value={2} />
        </RadioGroup>
      )
      // Select first input
      const firstRadio = container.querySelector(".knit-radio-input")
      expect(firstRadio).toBeInTheDocument()
      fireEvent.click(firstRadio)
      expect(onClick).toBeCalled()
      expect(onClick.mock.calls[0].length).toBe(1)
    })

    it("should call the provided onChange function", () => {
      const onChange = jest.fn()
      const { container } = render(
        <RadioGroup className="abc" onChange={onChange}>
          <RadioGroup.Item value={1}>Option 1</RadioGroup.Item>
          <RadioGroup.Item value={2} />
        </RadioGroup>
      )
      // select container element
      expect(container.querySelector(".abc")).toBeInTheDocument()
      fireEvent.click(container.querySelector(".abc"))
      expect(onChange).toBeCalled()
      expect(onChange.mock.calls[0].length).toBe(1)
    })
  })
})
