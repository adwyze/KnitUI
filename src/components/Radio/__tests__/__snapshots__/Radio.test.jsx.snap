// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio renders radio correctly Medium with 2 in a group 1`] = `
<DocumentFragment>
  .c0.knit-radio {
  white-space: nowrap;
  outline: none;
  display: inline-block;
  position: relative;
}

.c0.knit-radio + span {
  cursor: not-allowed;
  margin-left: 10px;
}

.c0.knit-radio.knit-radio-checked .knit-radio-inner::after {
  opacity: 1;
}

.c0.knit-radio.knit-radio-disabled .knit-radio-inner {
  cursor: not-allowed;
  opacity: 0.6;
  background-color: #f2f2f2;
}

.c0.knit-radio.knit-radio-disabled + span {
  color: #808080;
}

.c0.knit-radio.knit-radio-disabled .knit-radio-input {
  cursor: not-allowed;
}

.c2 {
  position: relative;
  top: 0;
  left: 0;
  display: inline-block;
  width: 1.4rem;
  height: 1.4rem;
  border-width: 1px;
  cursor: pointer;
  border-style: solid;
  border-radius: 14px;
  border-color: #002966;
  background-color: #FFFFFF;
}

.c2::after {
  position: absolute;
  width: 0.8rem;
  height: 0.8rem;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  border-radius: 0.7rem;
  border-top: 0;
  border-left: 0;
  content: " ";
  background-color: #002966;
  opacity: 0;
}

.c2:focus {
  box-shadow: 0px 0px 2px hsl(216,100%,20%);
  outline: none;
}

.c1 {
  position: absolute;
  left: 0;
  z-index: 9999;
  cursor: pointer;
  opacity: 0;
  top: 0;
  bottom: 0;
  right: 0;
}

<label
    style="display: flex;"
  >
    <span
      class="knit-radio  knit-radio-checked knit-radio-disabled c0"
      disabled=""
    >
      <input
        checked=""
        class="knit-radio-input c1"
        disabled=""
        name="nodejs"
        type="radio"
        value="1"
      />
      <span
        class="knit-radio-inner c2"
        disabled=""
      />
    </span>
    <span>
      Option 1
    </span>
  </label>
  .c0.knit-radio {
  white-space: nowrap;
  outline: none;
  display: inline-block;
  position: relative;
}

.c0.knit-radio + span {
  cursor: not-allowed;
  margin-left: 10px;
}

.c0.knit-radio.knit-radio-checked .knit-radio-inner::after {
  opacity: 1;
}

.c0.knit-radio.knit-radio-disabled .knit-radio-inner {
  cursor: not-allowed;
  opacity: 0.6;
  background-color: #f2f2f2;
}

.c0.knit-radio.knit-radio-disabled + span {
  color: #808080;
}

.c0.knit-radio.knit-radio-disabled .knit-radio-input {
  cursor: not-allowed;
}

.c2 {
  position: relative;
  top: 0;
  left: 0;
  display: inline-block;
  width: 1.4rem;
  height: 1.4rem;
  border-width: 1px;
  cursor: pointer;
  border-style: solid;
  border-radius: 14px;
  border-color: #002966;
  background-color: #FFFFFF;
}

.c2::after {
  position: absolute;
  width: 0.8rem;
  height: 0.8rem;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  border-radius: 0.7rem;
  border-top: 0;
  border-left: 0;
  content: " ";
  background-color: #002966;
  opacity: 0;
}

.c2:focus {
  box-shadow: 0px 0px 2px hsl(216,100%,20%);
  outline: none;
}

.c1 {
  position: absolute;
  left: 0;
  z-index: 9999;
  cursor: pointer;
  opacity: 0;
  top: 0;
  bottom: 0;
  right: 0;
}

<label
    style="display: flex;"
  >
    <span
      class="knit-radio   knit-radio-disabled c0"
      disabled=""
    >
      <input
        class="knit-radio-input c1"
        disabled=""
        name="nodejs"
        type="radio"
        value="2"
      />
      <span
        class="knit-radio-inner c2"
        disabled=""
      />
    </span>
    <span>
      Option 2
    </span>
  </label>
</DocumentFragment>
`;

exports[`Radio renders radio correctly Small with 2 in group 1`] = `
<DocumentFragment>
  .c0.knit-radio {
  white-space: nowrap;
  outline: none;
  display: inline-block;
  position: relative;
}

.c0.knit-radio + span {
  cursor: pointer;
  margin-left: 10px;
}

.c0.knit-radio.knit-radio-checked .knit-radio-inner::after {
  opacity: 1;
}

.c0.knit-radio.knit-radio-disabled .knit-radio-inner {
  cursor: not-allowed;
  opacity: 0.6;
  background-color: #f2f2f2;
}

.c0.knit-radio.knit-radio-disabled + span {
  color: #808080;
}

.c0.knit-radio.knit-radio-disabled .knit-radio-input {
  cursor: not-allowed;
}

.c2 {
  position: relative;
  top: 0;
  left: 0;
  display: inline-block;
  width: 1.2rem;
  height: 1.2rem;
  border-width: 1px;
  cursor: pointer;
  border-style: solid;
  border-radius: 14px;
  border-color: #002966;
  background-color: #FFFFFF;
}

.c2::after {
  position: absolute;
  width: 0.6rem;
  height: 0.6rem;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  border-radius: 0.7rem;
  border-top: 0;
  border-left: 0;
  content: " ";
  background-color: #002966;
  opacity: 0;
}

.c2:focus {
  box-shadow: 0px 0px 2px hsl(216,100%,20%);
  outline: none;
}

.c1 {
  position: absolute;
  left: 0;
  z-index: 9999;
  cursor: pointer;
  opacity: 0;
  top: 0;
  bottom: 0;
  right: 0;
}

<label
    style="display: flex;"
  >
    <span
      class="knit-radio    c0"
    >
      <input
        class="knit-radio-input c1"
        name="nodejs"
        type="radio"
        value="1"
      />
      <span
        class="knit-radio-inner c2"
      />
    </span>
    <span>
      Option 1
    </span>
  </label>
  .c0.knit-radio {
  white-space: nowrap;
  outline: none;
  display: inline-block;
  position: relative;
}

.c0.knit-radio + span {
  cursor: pointer;
  margin-left: 10px;
}

.c0.knit-radio.knit-radio-checked .knit-radio-inner::after {
  opacity: 1;
}

.c0.knit-radio.knit-radio-disabled .knit-radio-inner {
  cursor: not-allowed;
  opacity: 0.6;
  background-color: #f2f2f2;
}

.c0.knit-radio.knit-radio-disabled + span {
  color: #808080;
}

.c0.knit-radio.knit-radio-disabled .knit-radio-input {
  cursor: not-allowed;
}

.c2 {
  position: relative;
  top: 0;
  left: 0;
  display: inline-block;
  width: 1.2rem;
  height: 1.2rem;
  border-width: 1px;
  cursor: pointer;
  border-style: solid;
  border-radius: 14px;
  border-color: #002966;
  background-color: #FFFFFF;
}

.c2::after {
  position: absolute;
  width: 0.6rem;
  height: 0.6rem;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  border-radius: 0.7rem;
  border-top: 0;
  border-left: 0;
  content: " ";
  background-color: #002966;
  opacity: 0;
}

.c2:focus {
  box-shadow: 0px 0px 2px hsl(216,100%,20%);
  outline: none;
}

.c1 {
  position: absolute;
  left: 0;
  z-index: 9999;
  cursor: pointer;
  opacity: 0;
  top: 0;
  bottom: 0;
  right: 0;
}

<label
    style="display: flex;"
  >
    <span
      class="knit-radio  knit-radio-checked  c0"
    >
      <input
        class="knit-radio-input c1"
        name="nodejs"
        type="radio"
        value="2"
      />
      <span
        class="knit-radio-inner c2"
      />
    </span>
    <span>
      Option 2
    </span>
  </label>
</DocumentFragment>
`;

exports[`Radio renders radio correctly Small with 2 in group with disabled 1`] = `
<DocumentFragment>
  .c0.knit-radio {
  white-space: nowrap;
  outline: none;
  display: inline-block;
  position: relative;
}

.c0.knit-radio + span {
  cursor: not-allowed;
  margin-left: 10px;
}

.c0.knit-radio.knit-radio-checked .knit-radio-inner::after {
  opacity: 1;
}

.c0.knit-radio.knit-radio-disabled .knit-radio-inner {
  cursor: not-allowed;
  opacity: 0.6;
  background-color: #f2f2f2;
}

.c0.knit-radio.knit-radio-disabled + span {
  color: #808080;
}

.c0.knit-radio.knit-radio-disabled .knit-radio-input {
  cursor: not-allowed;
}

.c2 {
  position: relative;
  top: 0;
  left: 0;
  display: inline-block;
  width: 1.2rem;
  height: 1.2rem;
  border-width: 1px;
  cursor: pointer;
  border-style: solid;
  border-radius: 14px;
  border-color: #002966;
  background-color: #FFFFFF;
}

.c2::after {
  position: absolute;
  width: 0.6rem;
  height: 0.6rem;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  border-radius: 0.7rem;
  border-top: 0;
  border-left: 0;
  content: " ";
  background-color: #002966;
  opacity: 0;
}

.c2:focus {
  box-shadow: 0px 0px 2px hsl(216,100%,20%);
  outline: none;
}

.c1 {
  position: absolute;
  left: 0;
  z-index: 9999;
  cursor: pointer;
  opacity: 0;
  top: 0;
  bottom: 0;
  right: 0;
}

<label
    style="display: flex;"
  >
    <span
      class="knit-radio  knit-radio-checked knit-radio-disabled c0"
      disabled=""
    >
      <input
        checked=""
        class="knit-radio-input c1"
        disabled=""
        name="nodejs"
        type="radio"
        value="1"
      />
      <span
        class="knit-radio-inner c2"
        disabled=""
      />
    </span>
    <span>
      Option 1
    </span>
  </label>
  .c0.knit-radio {
  white-space: nowrap;
  outline: none;
  display: inline-block;
  position: relative;
}

.c0.knit-radio + span {
  cursor: not-allowed;
  margin-left: 10px;
}

.c0.knit-radio.knit-radio-checked .knit-radio-inner::after {
  opacity: 1;
}

.c0.knit-radio.knit-radio-disabled .knit-radio-inner {
  cursor: not-allowed;
  opacity: 0.6;
  background-color: #f2f2f2;
}

.c0.knit-radio.knit-radio-disabled + span {
  color: #808080;
}

.c0.knit-radio.knit-radio-disabled .knit-radio-input {
  cursor: not-allowed;
}

.c2 {
  position: relative;
  top: 0;
  left: 0;
  display: inline-block;
  width: 1.2rem;
  height: 1.2rem;
  border-width: 1px;
  cursor: pointer;
  border-style: solid;
  border-radius: 14px;
  border-color: #002966;
  background-color: #FFFFFF;
}

.c2::after {
  position: absolute;
  width: 0.6rem;
  height: 0.6rem;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  border-radius: 0.7rem;
  border-top: 0;
  border-left: 0;
  content: " ";
  background-color: #002966;
  opacity: 0;
}

.c2:focus {
  box-shadow: 0px 0px 2px hsl(216,100%,20%);
  outline: none;
}

.c1 {
  position: absolute;
  left: 0;
  z-index: 9999;
  cursor: pointer;
  opacity: 0;
  top: 0;
  bottom: 0;
  right: 0;
}

<label
    style="display: flex;"
  >
    <span
      class="knit-radio   knit-radio-disabled c0"
      disabled=""
    >
      <input
        class="knit-radio-input c1"
        disabled=""
        name="nodejs"
        type="radio"
        value="2"
      />
      <span
        class="knit-radio-inner c2"
        disabled=""
      />
    </span>
    <span>
      Option 2
    </span>
  </label>
</DocumentFragment>
`;

exports[`Radio renders radio correctly renders with no child correctly 1`] = `<DocumentFragment />`;

exports[`Radio renders radio correctly renders with single child correctly 1`] = `
<DocumentFragment>
  .c0.knit-radio {
  white-space: nowrap;
  outline: none;
  display: inline-block;
  position: relative;
}

.c0.knit-radio + span {
  cursor: pointer;
  margin-left: 10px;
}

.c0.knit-radio.knit-radio-checked .knit-radio-inner::after {
  opacity: 1;
}

.c0.knit-radio.knit-radio-disabled .knit-radio-inner {
  cursor: not-allowed;
  opacity: 0.6;
  background-color: #f2f2f2;
}

.c0.knit-radio.knit-radio-disabled + span {
  color: #808080;
}

.c0.knit-radio.knit-radio-disabled .knit-radio-input {
  cursor: not-allowed;
}

.c2 {
  position: relative;
  top: 0;
  left: 0;
  display: inline-block;
  width: 1.4rem;
  height: 1.4rem;
  border-width: 1px;
  cursor: pointer;
  border-style: solid;
  border-radius: 14px;
  border-color: #002966;
  background-color: #FFFFFF;
}

.c2::after {
  position: absolute;
  width: 0.8rem;
  height: 0.8rem;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  border-radius: 0.7rem;
  border-top: 0;
  border-left: 0;
  content: " ";
  background-color: #002966;
  opacity: 0;
}

.c2:focus {
  box-shadow: 0px 0px 2px hsl(216,100%,20%);
  outline: none;
}

.c1 {
  position: absolute;
  left: 0;
  z-index: 9999;
  cursor: pointer;
  opacity: 0;
  top: 0;
  bottom: 0;
  right: 0;
}

<label
    style="display: flex;"
  >
    <span
      class="knit-radio    c0"
    >
      <input
        class="knit-radio-input c1"
        name="nodejs"
        type="radio"
        value="1"
      />
      <span
        class="knit-radio-inner c2"
      />
    </span>
  </label>
</DocumentFragment>
`;
